README.md
# sensible-env-logger

[<img alt="github" src="https://img.shields.io/badge/github-source-8da0cb?style=for-the-badge&labelColor=555555&logo=github" height="25">](https://github.com/rnag/sensible-env-logger)
[<img alt="crates.io" src="https://img.shields.io/crates/v/sensible-env-logger.svg?style=for-the-badge&color=fc8d62&logo=rust" height="25">](https://crates.io/crates/sensible-env-logger)
[<img alt="docs.rs" src="https://img.shields.io/docsrs/sensible-env-logger/latest?style=for-the-badge&labelColor=555555&logoColor=white&logo=data:image/svg+xml;base64,PHN2ZyByb2xlPSJpbWciIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmlld0JveD0iMCAwIDUxMiA1MTIiPjxwYXRoIGZpbGw9IiNmNWY1ZjUiIGQ9Ik00ODguNiAyNTAuMkwzOTIgMjE0VjEwNS41YzAtMTUtOS4zLTI4LjQtMjMuNC0zMy43bC0xMDAtMzcuNWMtOC4xLTMuMS0xNy4xLTMuMS0yNS4zIDBsLTEwMCAzNy41Yy0xNC4xIDUuMy0yMy40IDE4LjctMjMuNCAzMy43VjIxNGwtOTYuNiAzNi4yQzkuMyAyNTUuNSAwIDI2OC45IDAgMjgzLjlWMzk0YzAgMTMuNiA3LjcgMjYuMSAxOS45IDMyLjJsMTAwIDUwYzEwLjEgNS4xIDIyLjEgNS4xIDMyLjIgMGwxMDMuOS01MiAxMDMuOSA1MmMxMC4xIDUuMSAyMi4xIDUuMSAzMi4yIDBsMTAwLTUwYzEyLjItNi4xIDE5LjktMTguNiAxOS45LTMyLjJWMjgzLjljMC0xNS05LjMtMjguNC0yMy40LTMzLjd6TTM1OCAyMTQuOGwtODUgMzEuOXYtNjguMmw4NS0zN3Y3My4zek0xNTQgMTA0LjFsMTAyLTM4LjIgMTAyIDM4LjJ2LjZsLTEwMiA0MS40LTEwMi00MS40di0uNnptODQgMjkxLjFsLTg1IDQyLjV2LTc5LjFsODUtMzguOHY3NS40em0wLTExMmwtMTAyIDQxLjQtMTAyLTQxLjR2LS42bDEwMi0zOC4yIDEwMiAzOC4ydi42em0yNDAgMTEybC04NSA0Mi41di03OS4xbDg1LTM4Ljh2NzUuNHptMC0xMTJsLTEwMiA0MS40LTEwMi00MS40di0uNmwxMDItMzguMiAxMDIgMzguMnYuNnoiPjwvcGF0aD48L3N2Zz4K" height="25">](https://docs.rs/sensible-env-logger)
[<img alt="build status" src="https://img.shields.io/github/actions/workflow/status/rnag/sensible-env-logger/build.yml?branch=main&style=for-the-badge" height="25">](https://github.com/rnag/sensible-env-logger/actions?query=branch%3Amain)

> A pretty, sensible logger for Rust - ideal for running *examples* and *tests* on a crate of choice.

This is a thin wrapper around [pretty_env_logger]. It requires minimal configuration to set up, and writes
to standard error with nice colored output for log levels.

![example-output](readme-example.png)

[pretty_env_logger]: https://docs.rs/pretty_env_logger

---

This crate works with Cargo with a `Cargo.toml` like:

```toml
[dependencies]
log = "0.4"
sensible-env-logger = "0.3"
```

## Getting started

The `sensible-env-logger` is ideally used when writing out [examples] 
or [tests] for a Cargo project or crate. It can be added as a
[`dev-dependency`] if needed.

[examples]: http://xion.io/post/code/rust-examples.html
[tests]: https://doc.rust-lang.org/book/ch11-01-writing-tests.html
[`dev-dependency`]: https://doc.rust-lang.org/rust-by-example/testing/dev_dependencies.html

First, add some usage to your application:

```rust
#[macro_use] extern crate log;

fn main() {
    sensible_env_logger::init!();

    info!("such information");
    warn!("o_O");
    error!("much error");
}
```

Then run your app with `cargo`, and you should see the full log output:

```console
cargo run
```

Alternatively, run your app with the environment variables that control the log level
for *external* crates as well as *your* crate explicitly set:

```console
GLOBAL_RUST_LOG=warn RUST_LOG=trace cargo run
```

Even though this crate has the name *env* in it, using the `sensible-env-logger`
in code is dead simple, and requires minimal configuration.

## Examples

You can check out sample usage of this crate in the [examples/](https://github.com/rnag/sensible-env-logger/tree/main/examples)
folder in the project repo on GitHub.

## Defaults

The helper macros below can be used to configure the global logger
with *sensible* defaults. Their sample log output is also shown.

> Note: any helper macros, such as `init!()`, should be called
> early in the execution of a Rust program.

### `init!()`

Initializes the global logger with a pretty, sensible env logger.

```console
INFO  my_module         > an informational message
```

### `init_timed!()`

Initializes the global logger with a *timed* pretty, sensible env logger.

```console
2022-03-24T17:15:31.683Z INFO  my_module         > an informational message
```

### `init_timed_short!()` - [`*`]

Initializes the global logger with a *localized time* pretty, sensible env logger.

```console
12:15:31.683 INFO  my_module         > an informational message
```

### `init_timed_local!()` - [`*`]

Initializes the global logger with a "no-frills" local date/time
pretty, sensible env logger.

```console
2022-03-24 12:15:31.683 - INFO  my_module         > an informational message
```

### `init_timed_local_iso!()` - [`*`]

Initializes the global logger with a local-timed pretty, sensible
env logger.

> This variant formats log messages with a localized timestamp,
> in ISO-8601/ RFC 3339 date & time format.

```console
2022-03-24T12:15:31.683+08:00 INFO  my_module         > an informational message
```

## Optional Features

[`*`]: #local-time

### Local Time

Using the macros marked with a `*` above, require the `local-time` feature to be enabled:

```toml
[dev-dependencies]
sensible-env-logger = { version = "0.3", features = ["local-time"] }
```

## In Tests

When running *tests* on a crate, you can use the `safe_init!()` macro
as shown below. This should ignore errors when initializing the global
logger multiple times.

```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_one() {
        sensible_env_logger::safe_init!();
        trace!("hello from the first test")
    }

    #[test]
    fn test_two() {
        sensible_env_logger::safe_init!();
        trace!("hello from the second test")
    }
}
```

## Rationale

Imagine you are testing out a Cargo project named `my_rust_project`. That is,
the `Cargo.toml` in your project would look something like this:

```toml,no_sync
[package]
name = "my-rust-project"

[dependencies]
log = "*"
```

Assuming you are building a library, your `src/lib.rs` could look like this:

```rust
#[macro_use] extern crate log;

pub fn my_awesome_fn() {
    trace!("getting ready to do something cool...");
    std::thread::sleep(std::time::Duration::from_millis(500));
    info!("finished!");
}
```

You then create a new [example] file named `examples/my_example.rs`, with the following
contents:

```rust
use my_rust_project::my_awesome_fn;
#[macro_use] extern crate log;

fn main() {
    debug!("my debug message");
    my_awesome_fn();
    error!("oops! something went wrong!");
}
```

[example]: http://xion.io/post/code/rust-examples.html

You can run the new file with `cargo run --example my_example`. The problem here is that you won't
get any terminal output by default; this is because you need to set up the `RUST_LOG`
environment variable beforehand, in order to see the expected log output.

Setting the `RUST_LOG` variable works, but there are a few issues with this. For example, what if your Cargo project uses other external libraries? Ideally you want to
see the `trace` logs from your own project (the crate under test), but *not* the
`trace` logs from these other libraries. In that case, setting `RUST_LOG=trace`
doesn't seem the best approach here.

You could then set the `RUST_LOG` environment variable to the following [log format]:

```shell
$ export RUST_LOG='warning,my_rust_project=trace,my_example=trace'
```

When leveraging the [pretty_env_logger] crate and adding a `pretty_env_logger::init()`
at the top of the `main` function, this does now work as expected and produce the desired log output.

However, there are few limitations with this approach:

* In your Cargo project, you'd need to update the documentation for running examples to
  mention that you need to export the `RUST_LOG` env variable explicitly. For instance,
  you'd need to mention that an example is ideally run like `RUST_LOG=trace cargo run --example my_example`.

* You'd need to remember to set the `RUST_LOG` env variable each time. This can be troublesome
  when your Windows machine reboots for example, or whenever you open a new terminal window.

To solve these minor issues you can simply use the `sensible_env_logger` crate, which
automatically sets up *sensible* defaults; this involves generating and using a
directive string in the same form as the `$RUST_LOG` environment variable.

Now, the updated code in the `examples/my_example.rs` would look like this:

```rust
use my_rust_project::my_awesome_fn;
#[macro_use] extern crate log;

fn main() {
    sensible_env_logger::init!();

    debug!("my debug message");
    my_awesome_fn();
    error!("oops! something went wrong!");
}
```

[log format]: https://rust-lang-nursery.github.io/rust-cookbook/development_tools/debugging/config_log.html

## Contributing

Contributions are welcome! Open a pull request to fix a bug, or [open an issue][]
to discuss a new feature or change.

Check out the [Contributing][] section in the docs for more info.

[Contributing]: CONTRIBUTING.md
[open an issue]: https://github.com/rnag/sensible-env-logger/issues

## License

This project is proudly licensed under the MIT license ([LICENSE](LICENSE)
or http://opensource.org/licenses/MIT).

`sensible-env-logger` can be distributed according to the MIT license. Contributions
will be accepted under the same license.

## Authors

* [Ritvik Nag](https://github.com/rnag)


CHANGELOG.md
# Changelog

This project follows semantic versioning.

Possible header types:

- `Features` for any new features added, or for backwards-compatible
  changes to existing functionality.
- `Bug Fixes` for any bug fixes.
- `Breaking Changes` for any backwards-incompatible changes.

## [Unreleased]
<!--
### Features
- Added a new struct `MyStruct` with the following methods:
  - `my_method()`
  - `other_method()`
-->

## v0.3.2 (2023-02-05)

### Bug Fixes

* Fix readme badge, as per [badges/shields#8671](https://github.com/badges/shields/issues/8671).

## v0.3.1 (2022-08-10)

### Bug Fixes

* Correct example usage and log format as it appears in the documentation.

## v0.3.0 (2022-08-09)

### Features

- Add new helper macros to log the local date
  and local date/time ISO:
  - `init_timed_local` / `safe_init_timed_local` / `try_init_timed_local`
  - `init_timed_local_iso` / `safe_init_timed_local_iso` / `try_init_timed_local_iso`
- Add new examples `log_timed_local` and `log_timed_local_iso`
  to confirm intended functionality.

## v0.2.0 (2022-03-19)

### Features

- Add helper macros `safe_init`, `safe_init_timed`, and `safe_init_timed_short`.
  This should make it easier to run *tests* on a crate of choice.

### Bug Fixes

- Fix so module name is correctly generated.
  - In some edge cases, mainly in *tests* within binaries such `src/bin/my_bin/my_file.rs`,
    it only enabled logs on the nested module path, ex. `my_bin::my_module::tests`. Now
    it should instead enable logs on the base module, `my_bin` in this example.

## v0.1.0 (2022-03-13)

- Initial Release on [crates.io] :tada:

[crates.io]: https://crates.io/crates/sensible-env-logger


src/lib.rs
#![doc(html_root_url = "https://docs.rs/sensible-env-logger/0.3.2")]
#![warn(rust_2018_idioms, missing_docs)]
#![deny(warnings, dead_code, unused_imports, unused_mut)]
#![allow(clippy::uninlined_format_args)]
//! [![github]](https://github.com/rnag/sensible-env-logger)&ensp;[![crates-io]](https://crates.io/crates/sensible-env-logger)&ensp;[![docs-rs]](https://docs.rs/sensible-env-logger)
//!
//! [github]: https://img.shields.io/badge/github-8da0cb?style=for-the-badge&labelColor=555555&logo=github
//! [crates-io]: https://img.shields.io/badge/crates.io-fc8d62?style=for-the-badge&labelColor=555555&logo=rust
//! [docs-rs]: https://img.shields.io/badge/docs.rs-66c2a5?style=for-the-badge&labelColor=555555&logoColor=white&logo=data:image/svg+xml;base64,PHN2ZyByb2xlPSJpbWciIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmlld0JveD0iMCAwIDUxMiA1MTIiPjxwYXRoIGZpbGw9IiNmNWY1ZjUiIGQ9Ik00ODguNiAyNTAuMkwzOTIgMjE0VjEwNS41YzAtMTUtOS4zLTI4LjQtMjMuNC0zMy43bC0xMDAtMzcuNWMtOC4xLTMuMS0xNy4xLTMuMS0yNS4zIDBsLTEwMCAzNy41Yy0xNC4xIDUuMy0yMy40IDE4LjctMjMuNCAzMy43VjIxNGwtOTYuNiAzNi4yQzkuMyAyNTUuNSAwIDI2OC45IDAgMjgzLjlWMzk0YzAgMTMuNiA3LjcgMjYuMSAxOS45IDMyLjJsMTAwIDUwYzEwLjEgNS4xIDIyLjEgNS4xIDMyLjIgMGwxMDMuOS01MiAxMDMuOSA1MmMxMC4xIDUuMSAyMi4xIDUuMSAzMi4yIDBsMTAwLTUwYzEyLjItNi4xIDE5LjktMTguNiAxOS45LTMyLjJWMjgzLjljMC0xNS05LjMtMjguNC0yMy40LTMzLjd6TTM1OCAyMTQuOGwtODUgMzEuOXYtNjguMmw4NS0zN3Y3My4zek0xNTQgMTA0LjFsMTAyLTM4LjIgMTAyIDM4LjJ2LjZsLTEwMiA0MS40LTEwMi00MS40di0uNnptODQgMjkxLjFsLTg1IDQyLjV2LTc5LjFsODUtMzguOHY3NS40em0wLTExMmwtMTAyIDQxLjQtMTAyLTQxLjR2LS42bDEwMi0zOC4yIDEwMiAzOC4ydi42em0yNDAgMTEybC04NSA0Mi41di03OS4xbDg1LTM4Ljh2NzUuNHptMC0xMTJsLTEwMiA0MS40LTEwMi00MS40di0uNmwxMDItMzguMiAxMDIgMzguMnYuNnoiPjwvcGF0aD48L3N2Zz4K
//!
//! <br>
//!
//! A simple logger, optionally configured via environment variables which
//! writes to standard error with nice colored output for log levels.
//! It sets up logging with "sensible" defaults that make it ideal for
//! running *[examples]* and *[tests]* on a crate of choice.
//!
//! [examples]: http://xion.io/post/code/rust-examples.html
//! [tests]: https://doc.rust-lang.org/book/ch11-01-writing-tests.html
//! <br>
//!
//! ## Usage
//!
//! Even though it has `env` in the name, the `sensible-env-logger`
//! requires minimal configuration and setup to use:
//!
//! ```
//! #[macro_use] extern crate log;
//!
//! fn main() {
//!     sensible_env_logger::init!();
//!
//!     trace!("a trace example");
//!     debug!("deboogging");
//!     info!("such information");
//!     warn!("o_O");
//!     error!("boom");
//! }
//! ```
//!
//! Run the program and you should see all the log output for your crate.
//!
//! Alternatively, run the program with the environment variables that control
//! the log level for *your* crate as well as *external* crates explicitly set,
//! like `RUST_LOG=debug` and `GLOBAL_RUST_LOG=error`.
//!
//! ## Defaults
//!
//! The defaults can be setup by calling `init!()` or `try_init!()` at the start
//! of the program.
//!
//! ## Examples
//!
//! You can check out sample usage of this crate in the [examples/](https://github.com/rnag/sensible-env-logger/tree/main/examples)
//! folder in the project repo on GitHub.
//!
//! ## Readme Docs
//!
//! You can find the crate's readme documentation on the
//! [crates.io] page, or alternatively in the [`README.md`] file on the GitHub project repo.
//!
//! [crates.io]: https://crates.io/crates/sensible-env-logger
//! [`README.md`]: https://github.com/rnag/sensible-env-logger
//!
//! ## Enable logging
//!
//! This crate uses [pretty_env_logger] and [env_logger] internally, so the
//! same ways of enabling logs through an environment variable are supported.
//!
//! The `sensible_env_logger` crate re-exports these crates, through the
//! `pretty` and `env` namespaces respectively.
//!
//! [pretty_env_logger]: https://docs.rs/pretty_env_logger
//! [env_logger]: https://docs.rs/env_logger

use std::borrow::Cow;

use env::Builder;
use log::{SetLoggerError, trace};
#[doc(hidden)]
pub use pretty_env_logger as pretty;
#[doc(hidden)]
pub use pretty_env_logger::env_logger as env;

#[cfg(feature = "local-time")]
pub use local_time::*;

/// Default log level for the Cargo crate or package under test.
pub(crate) const CRATE_LOG_LEVEL: &str = "trace";

/// Default log level for external crates, other than the one under test.
pub(crate) const GLOBAL_LOG_LEVEL: &str = "warn";

/// Initializes the global logger with a pretty, sensible env logger.
///
/// This should be called early in the execution of a Rust program, and the
/// global logger may only be initialized once. Future initialization attempts
/// will return an error.
///
/// # Panics
///
/// This macro fails to set the global logger if one has already been set.
#[macro_export]
macro_rules! init {
    () => {
        $crate::try_init!().unwrap()
    };
}

/// Initializes the global logger with a timed pretty, sensible env logger.
///
/// This should be called early in the execution of a Rust program, and the
/// global logger may only be initialized once. Future initialization attempts
/// will return an error.
///
/// # Panics
///
/// This macro fails to set the global logger if one has already been set.
#[macro_export]
macro_rules! init_timed {
    () => {
        $crate::try_init_timed!().unwrap();
    };
}

/// Initializes the global logger with a pretty, sensible env logger.
///
/// This variant should ideally only be used in **tests**. It should be called
/// early in the execution of a Rust program.
///
/// Future initialization attempts will *safely ignore* any errors.
#[macro_export]
macro_rules! safe_init {
    () => {
        let _ = $crate::try_init!();
    };
}

/// Initializes the global logger with a timed pretty, sensible env logger.
///
/// This variant should ideally only be used in **tests**. It should be called
/// early in the execution of a Rust program.
///
/// Future initialization attempts will *safely ignore* any errors.
#[macro_export]
macro_rules! safe_init_timed {
    () => {
        let _ = $crate::try_init_timed!();
    };
}

/// Initializes the global logger with a pretty, sensible env logger.
///
/// This should be called early in the execution of a Rust program, and the
/// global logger may only be initialized once. Future initialization attempts
/// will return an error.
///
/// # Errors
///
/// This macro fails to set the global logger if one has already been set.
#[macro_export]
macro_rules! try_init {
    () => {
        $crate::try_init_custom_env_and_builder(
            "RUST_LOG",
            "GLOBAL_RUST_LOG",
            env!("CARGO_PKG_NAME"),
            module_path!(),
            $crate::pretty::formatted_builder,
        )
    };
}

/// Initializes the global logger with a timed pretty, sensible env logger.
///
/// This should be called early in the execution of a Rust program, and the
/// global logger may only be initialized once. Future initialization attempts
/// will return an error.
///
/// # Errors
///
/// This macro fails to set the global logger if one has already been set.
#[macro_export]
macro_rules! try_init_timed {
    () => {
        $crate::try_init_custom_env_and_builder(
            "RUST_LOG",
            "GLOBAL_RUST_LOG",
            env!("CARGO_PKG_NAME"),
            module_path!(),
            $crate::pretty::formatted_timed_builder,
        )
    };
}

/// Initializes the global logger with a pretty, sensible env logger, with custom
/// variable names and a custom builder function.
///
/// This should be called early in the execution of a Rust program, and the
/// global logger may only be initialized once. Future initialization attempts
/// will return an error.
///
/// # Example
/// ```rust
/// let _ = sensible_env_logger::try_init_custom_env_and_builder(
///     "MY_RUST_LOG",
///     "MY_GLOBAL_RUST_LOG",
///     env!("CARGO_PKG_NAME"),
///     module_path!(),
///     sensible_env_logger::pretty::formatted_timed_builder,
/// );
/// ```
///
/// # How It works
///
/// The `package_name` and `module_name` arguments are ideally evaluated from
/// the `$CARGO_PKG_NAME` and `$CARGO_CRATE_NAME` environment variables
/// respectively. These environment variables are automatically set
/// by Cargo when compiling your crate. It then builds a custom directives
/// string in the same form as the `$RUST_LOG` environment variable, and then
/// parses this generated directives string using
/// `env_logger::Builder::parse_filters`.
///
/// # Errors
///
/// This function fails to set the global logger if one has already been set.
pub fn try_init_custom_env_and_builder(
    log_env_var: &str,
    global_log_env_var: &str,
    package_name: &str,
    module_name: &str,
    builder_fn: impl Fn() -> Builder,
) -> Result<(), SetLoggerError> {
    let package_name = package_name.replace('-', "_");
    let module_name = base_module(module_name);

    let log_level = get_env(log_env_var, CRATE_LOG_LEVEL);
    let global_log_level = get_env(global_log_env_var, GLOBAL_LOG_LEVEL);

    let filters_str = if log_level.contains('=') {
        // The env variable `$RUST_LOG` is set to a more complex value such as
        // `warn,my_module=info`. In that case, just pass through the value.
        log_level.into_owned()
    } else if package_name != module_name {
        format!(
            "{default_lvl},{pkg}={lvl},{mod}={lvl}",
            default_lvl = global_log_level,
            pkg = package_name,
            mod = module_name,
            lvl = log_level
        )
    } else {
        format!(
            "{default_lvl},{pkg}={lvl}",
            default_lvl = global_log_level,
            pkg = package_name,
            lvl = log_level
        )
    };

    let mut builder: Builder = builder_fn();
    builder.parse_filters(&filters_str);

    let result = builder.try_init();

    trace!("Filter: {}", filters_str);

    result
}

/// Retrieve the value of an environment variable.
pub(crate) fn get_env<'a>(env_var_name: &'a str, default: &'a str) -> Cow<'a, str> {
    match std::env::var(env_var_name) {
        Ok(value) => Cow::Owned(value),
        _ => Cow::Borrowed(default),
    }
}

/// Returns the base module name, given the path to a module.
///
/// # Example
/// ```no_test
/// assert_eq!(base_module("my_bin::my_module::tests"), "my_bin");
/// ```
///
pub(crate) fn base_module(module_name: &str) -> &str {
    match module_name.split_once("::") {
        Some((first, _)) => first,
        None => module_name,
    }
}

#[cfg(feature = "local-time")]
mod local_time {
    use std::fmt;
    use std::sync::atomic::{AtomicUsize, Ordering};

    use chrono::Local;
    use env::fmt::{Color, Style, StyledValue};
    use log::Level;

    use super::*;

    /// Local time zone format (only time)
    ///
    /// # Example
    /// `10:17:52.831`
    ///
    pub const TIME_ONLY_FMT: &str = "%l:%M:%S.%3f";

    /// Local time zone format
    ///
    /// # Example
    /// `2022-07-27 10:17:52.831 -`
    ///
    pub const LOCAL_TIME_FMT: &str = "%Y-%m-%d %l:%M:%S.%3f -";

    /// ISO 8601 / RFC 3339 date & time format.
    ///
    /// # Example
    /// `2022-07-27T17:34:44.531+08:00`
    ///
    pub const ISO_FMT: &str = "%Y-%m-%dT%H:%M:%S.%3f%:z";

    /// Initializes the global logger with an "abbreviated" timed pretty, sensible
    /// env logger.
    ///
    /// This should be called early in the execution of a Rust program, and the
    /// global logger may only be initialized once. Future initialization attempts
    /// will return an error.
    ///
    /// # Details
    ///
    /// This variant formats log messages with a localized timestamp, without
    /// the date part.
    ///
    /// ## Example
    ///
    /// ```console
    /// 12:15:31.683 INFO  my_module         > an info message!
    /// ```
    ///
    /// # Requirements
    ///
    /// Using this macro requires the `local-time` feature to be enabled:
    ///
    /// ```toml
    /// [dependencies]
    /// sensible-env-logger = { version = "*", features = ["local-time"] }
    /// ```
    ///
    /// # Panics
    ///
    /// This macro fails to set the global logger if one has already been set.
    #[macro_export]
    macro_rules! init_timed_short {
        () => {
            $crate::try_init_timed_short!().unwrap();
        };
    }

    /// Initializes the global logger with a "no-frills" local date/time
    /// pretty, sensible env logger.
    ///
    /// This should be called early in the execution of a Rust program, and the
    /// global logger may only be initialized once. Future initialization attempts
    /// will return an error.
    ///
    /// # Details
    ///
    /// This variant formats log messages with a localized timestamp,
    /// prefixed by the date part.
    ///
    /// ## Example
    ///
    /// ```console
    /// 2021-10-27 12:15:31.683 - INFO  my_module         > an info message!
    /// ```
    ///
    /// # Requirements
    ///
    /// Using this macro requires the `local-time` feature to be enabled:
    ///
    /// ```toml
    /// [dependencies]
    /// sensible-env-logger = { version = "*", features = ["local-time"] }
    /// ```
    ///
    /// # Panics
    ///
    /// This macro fails to set the global logger if one has already been set.
    #[macro_export]
    macro_rules! init_timed_local {
        () => {
            $crate::try_init_timed_local!().unwrap();
        };
    }

    /// Initializes the global logger with a local-timed pretty, sensible
    /// env logger.
    ///
    /// This should be called early in the execution of a Rust program, and the
    /// global logger may only be initialized once. Future initialization attempts
    /// will return an error.
    ///
    /// # Details
    ///
    /// This variant formats log messages with a localized timestamp and zone,
    /// in complete ISO-8601/ RFC 3339 date & time format.
    ///
    /// ## Example
    ///
    /// ```console
    /// 2022-10-27T12:15:31.683+08:00 INFO  my_module         > an info message!
    /// ```
    ///
    /// # Requirements
    ///
    /// Using this macro requires the `local-time` feature to be enabled:
    ///
    /// ```toml
    /// [dependencies]
    /// sensible-env-logger = { version = "*", features = ["local-time"] }
    /// ```
    ///
    /// # Panics
    ///
    /// This macro fails to set the global logger if one has already been set.
    #[macro_export]
    macro_rules! init_timed_local_iso {
        () => {
            $crate::try_init_timed_local_iso!().unwrap();
        };
    }

    /// Initializes the global logger with an "abbreviated" timed pretty, sensible
    /// env logger.
    /// See [`init_timed_short!`](macro.init_timed_short.html) for more info.
    ///
    /// This variant should ideally only be used in **tests**. It should be called
    /// early in the execution of a Rust program.
    ///
    /// Future initialization attempts will *safely ignore* any errors.
    #[macro_export]
    macro_rules! safe_init_timed_short {
        () => {
            let _ = $crate::try_init_timed_short!();
        };
    }

    /// Initializes the global logger with a "no-frills" local date/time
    /// pretty, sensible env logger.
    /// See [`init_timed_local!`](macro.init_timed_local.html) for more info.
    ///
    /// This variant should ideally only be used in **tests**. It should be called
    /// early in the execution of a Rust program.
    ///
    /// Future initialization attempts will *safely ignore* any errors.
    #[macro_export]
    macro_rules! safe_init_timed_local {
        () => {
            let _ = $crate::try_init_timed_local!();
        };
    }

    /// Initializes the global logger with a local-timed pretty, sensible
    /// env logger.
    /// See [`init_timed_local_iso!`](macro.init_timed_local_iso.html) for more info.
    ///
    /// This variant should ideally only be used in **tests**. It should be called
    /// early in the execution of a Rust program.
    ///
    /// Future initialization attempts will *safely ignore* any errors.
    #[macro_export]
    macro_rules! safe_init_timed_local_iso {
        () => {
            let _ = $crate::try_init_timed_local_iso!();
        };
    }

    /// Initializes the global logger with an "abbreviated" timed pretty, sensible
    /// env logger.
    ///
    /// See [`init_timed_short!`](macro.init_timed_short.html) for more info.
    ///
    /// This should be called early in the execution of a Rust program, and the
    /// global logger may only be initialized once. Future initialization attempts
    /// will return an error.
    ///
    /// # Errors
    ///
    /// This macro fails to set the global logger if one has already been set.
    #[macro_export]
    macro_rules! try_init_timed_short {
        () => {
            $crate::try_init_custom_env_and_builder(
                "RUST_LOG",
                "GLOBAL_RUST_LOG",
                env!("CARGO_PKG_NAME"),
                module_path!(),
                $crate::formatted_local_time_builder_fn($crate::TIME_ONLY_FMT),
            )
        };
    }

    /// Initializes the global logger with a "no-frills" local date/time
    /// pretty, sensible env logger.
    ///
    /// See [`init_timed_local!`](macro.init_timed_local.html) for more info.
    ///
    /// This should be called early in the execution of a Rust program, and the
    /// global logger may only be initialized once. Future initialization attempts
    /// will return an error.
    ///
    /// # Errors
    ///
    /// This macro fails to set the global logger if one has already been set.
    #[macro_export]
    macro_rules! try_init_timed_local {
        () => {
            $crate::try_init_custom_env_and_builder(
                "RUST_LOG",
                "GLOBAL_RUST_LOG",
                env!("CARGO_PKG_NAME"),
                module_path!(),
                $crate::formatted_local_time_builder_fn($crate::LOCAL_TIME_FMT),
            )
        };
    }

    /// Initializes the global logger with a local-timed pretty, sensible
    /// env logger.
    ///
    /// See [`init_timed_local_iso!`](macro.init_timed_local_iso.html) for more info.
    ///
    /// This should be called early in the execution of a Rust program, and the
    /// global logger may only be initialized once. Future initialization attempts
    /// will return an error.
    ///
    /// # Errors
    ///
    /// This macro fails to set the global logger if one has already been set.
    #[macro_export]
    macro_rules! try_init_timed_local_iso {
        () => {
            $crate::try_init_custom_env_and_builder(
                "RUST_LOG",
                "GLOBAL_RUST_LOG",
                env!("CARGO_PKG_NAME"),
                module_path!(),
                $crate::formatted_local_time_builder_fn($crate::ISO_FMT),
            )
        };
    }

    /// Returns a function (closure) that returns a formatted builder which
    /// adds local time to log messages, per a specified date/time format.
    ///
    /// ## Example
    /// ```console
    /// 12:15:31.683 INFO  my_module         > an info message!
    /// ```
    ///
    pub fn formatted_local_time_builder_fn(fmt: &'static str) -> impl Fn() -> Builder {
        || {
            let mut builder = Builder::new();

            builder.format(|f, record| {
                use std::io::Write;
                let target = record.target();
                let max_width = max_target_width(target);

                let mut style = f.style();
                let level = colored_level(&mut style, record.level());

                let mut style = f.style();
                let target = style.set_bold(true).value(Padded {
                    value: target,
                    width: max_width,
                });

                let time = Local::now().format(fmt);

                writeln!(f, " {} {} {} > {}", time, level, target, record.args(),)
            });

            builder
        }
    }

    /// Helper functions
    ///
    /// Below are copied verbatim from [`pretty_env_logger`]
    ///
    /// [`pretty_env_logger`]: https://github.com/seanmonstar/pretty-env-logger/blob/master/src/lib.rs
    ///

    struct Padded<T> {
        value: T,
        width: usize,
    }

    impl<T: fmt::Display> fmt::Display for Padded<T> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            write!(f, "{: <width$}", self.value, width = self.width)
        }
    }

    static MAX_MODULE_WIDTH: AtomicUsize = AtomicUsize::new(0);

    fn max_target_width(target: &str) -> usize {
        let max_width = MAX_MODULE_WIDTH.load(Ordering::Relaxed);
        if max_width < target.len() {
            MAX_MODULE_WIDTH.store(target.len(), Ordering::Relaxed);
            target.len()
        } else {
            max_width
        }
    }

    fn colored_level(style: &'_ mut Style, level: Level) -> StyledValue<'_, &'static str> {
        match level {
            Level::Trace => style.set_color(Color::Magenta).value("TRACE"),
            Level::Debug => style.set_color(Color::Blue).value("DEBUG"),
            Level::Info => style.set_color(Color::Green).value("INFO "),
            Level::Warn => style.set_color(Color::Yellow).value("WARN "),
            Level::Error => style.set_color(Color::Red).value("ERROR"),
        }
    }
}

#[cfg(test)]
mod tests {
    use log::*;

    use super::*;

    #[test]
    fn logging_in_tests() {
        // Initialize the global logger with sensible defaults
        init!();

        trace!("A simple trace message");
        debug!("Debugging something...");
        warn!("This is a WARNING!");
    }

    #[test]
    fn test_base_module_simple() {
        let result = base_module("hello_world");
        assert_eq!(result, "hello_world");
    }

    #[test]
    fn test_base_module_with_nested_path() {
        let result = base_module("my_bin::my_module::tests");
        assert_eq!(result, "my_bin");
    }
}
